src/clj/commiteth/util/crypto_fiat_value.clj

```diff
@@ -1,66 +1,100 @@
-        data (json-api-request url)]
-    (-> (first data)
-        (get "price_usd")
-        (read-string))))
-
-(defn- get-price-fn []
-  (let [fns {:cryptonator get-token-usd-price-cryptonator
-                      :coinmarketcap get-token-usd-price-coinmarketcap}]
-    (get fns (fiat-api-provider))))
+  [tla token-id]
+  {:pre [(some? token-id)]}
+  (log/infof "%s: Getting price-data from CoinMarketCap (token-id %s)" tla token-id)
+  (-> (json-api-request (format "https://api.coinmarketcap.com/v1/ticker/%s" token-id))
+      (first)
+      (get "price_usd")
+      (edn/read-string)))
 
-(defn bounty-usd-value
-  "Get current USD value of a bounty. bounty is a map of token-tla (keyword) to value"
-  [bounty]
-  (let [get-token-usd-price (get-price-fn)]
-    (reduce + (map (fn [[tla value]]
-                     (let [usd-price (get-token-usd-price tla)]
-                       (* usd-price value)))
-                   bounty))))
+(defrecord PriceInfo [tla usd date])
 
+(defn recent? [price-info]
+  (t/after? (:date price-info) (t/minus (t/now) (t/minutes 5))))
+
+(defprotocol IFiatCryptoConverter
+  (start [_])
+  (convert-usd [_ tla amount]))
+
+(defrecord FiatCryptoConverter [provider state]
+  IFiatCryptoConverter
+  (start [_]
+    (when (= :coinmarketcap provider)
+      (swap! state assoc :tla->id (make-tla-to-id-mapping))))
+  (convert-usd [_ tla amount]
+    (if-let [recent-price-info (and (some-> (get-in @state [:price-info tla]) recent?)
+                                    ;; return actual price info
+                                    (get-in @state [:price-info tla]))]
+      (* (:usd recent-price-info) amount)
+      ;; if we don't have price-info we need to fetch & store it
+      (let [price (case provider
+                    :coinmarketcap (get-token-usd-price-coinmarketcap
+                                    tla
+                                    (get-in @state [:tla->id tla]))
+                    :cryptonator (get-token-usd-price-cryptonator tla))]
+        (swap! state assoc-in [:price-info tla] (->PriceInfo tla price (t/now)))
+        (* price amount)))))
 
 (mount/defstate
-  crypto-fiat-util
-  :start
-  (do
-    (reset! tla-to-id-mapping (make-tla-to-id-mapping))
-    (log/info "crypto-fiat-util started"))
-  :stop
-  (log/info "crypto-fiat-util stopped"))
+  fiat-converter
+  :start (let [provider (fiat-api-provider)]
+           (log/infof "Starting FiatCryptoConverter %s" provider)
+           (doto (->FiatCryptoConverter provider (atom {}))
+             (start)))
+  :stop  (log/info "Stopping FiatCryptoConverter"))
+
+;; public api ------------------------------------------------------------------
+
+(defn bounty-usd-value
+  "Get current USD value for the crypto-amounts passed as argument.
+  Example: {:ETH 123, :SNT 456}"
+  [crypto-amounts]
+  (->> crypto-amounts
+       (map (fn [[tla value]]
+              (convert-usd fiat-converter tla value)))
+       (reduce + 0)))
+
+
+(comment
```

Comment:
had been meaning to ask about the use of these. are they meant to indicate how one might easily test this functionality if the comment was removed?

Reply:
I often use these to keep REPL snippets related to that namespace around. Usually I move my curser after the form I want to evaluate and evaluate the preceding form (`C-x C-e` in Emacs I think). This way I can easily run smaller functions and figure out if they're doing what they're supposed to :)

Reply:
Makes total sense! Thanks for sharing.